torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1308,9 +1308,11 @@ def call_function(
         from torch.utils.checkpoint import noop_context_fn
         from .builder import wrap_fx_proxy
 
+        context_fn_id = None
         if "context_fn" in kwargs and kwargs["context_fn"] != noop_context_fn:
             context_fn = kwargs.pop("context_fn")
-            self.value.context_fn = context_fn.fn
+            self.value.context_fn_list.append(context_fn.fn)
+            context_fn_id = len(self.value.context_fn_list) - 1
```

Comment:
I don't like this. It's a good way to have memory leak bugs. 

Prefer if you can use a hash table (rely on id of wrapper object) and clean up after the lifetime of the higher order op is complete.

Reply:
Please see updated PR - we don't have a good concept of "lifetime of the higher order op", so I attach the context func to the graph module itself. When the graph module is gc'ed, the context func will be gc'ed too.

Reply:
Hmm, seems better now! 

Can't be completely sure that no two graphmodule can be checkpointed twice? But it seems quite safe as we only add the `_checkpoint_context_fn` meta to the GM that we ourselves create, correct?

It seems like the right object to add this data too. Do you agree @zou3519 ?

Reply:
There were some comments about serializing the graph, can the GM meta be serialized? @Chillee